### ISSUE TITLE
Please add a repository_ctx method for reading files by Label

### ISSUE THREAD

**Issue by jmillikin** [bazelbuild/bazel#3766]
A `repository_ctx` can have Label attributes that point to files in other repositories (useful for tool plumbing), but such attributes are awkward to work with because there's no direct way to read their content.

The best workaround I've found is to use symlinks and run `cat` as a subprocess, which certainly deserves an entry in the [Useless Use of Cat Award](http://porkmail.org/era/unix/award.html).

```python
def _some_repo_rule(ctx):
  ctx.symlink(ctx.attr.genrule_cmd_file, "external-genrule-cmd.sh")
  cat_genrule_cmd = ctx.execute(["cat", "external-genrule-cmd.sh"])
  if cat_genrule_cmd.return_code != 0:
    fail("Failed to read genrule command %r: %s" % (
        ctx.attr.genrule_cmd_file, cat_genrule_cmd.stderr))
  genrule_content = cat_genrule_cmd.stdout
```

That whole mess could, ideally, be simplified down to this:

```python
def _some_repo_rule(ctx):
  genrule_content = ctx.read_file(ctx.attr.genrule_cmd_file)
```

**Comment by ittaiz** [comment]
This sounds related to (though definitely not a duplicate of) #3438

**Comment by jmillikin** [comment]
PR: https://github.com/bazelbuild/bazel/pull/7309

**Comment by dslomov** [comment]
What is the motivation for reading the file into Starlark rule?

**Comment by cgruber** [comment]
Getting metadata in that you can use to generate the repository more effectively.  In my case, for example, getting teh appropriate maven metadata (pom files) to ensure that we generate dependencies properly in the generated bazel targets.  Right now I'm using "cat" which doesn't work on windows.  I don't want to have to set up a per-platform toolchain just to read a file.

**Comment by jmillikin** [comment]
I've been read files in repository rules for two reasons:

1. Sometimes the choice of which artifacts to download is driven by a file from the main workspace. For example, Python or Node.JS projects save their download URLs to "lock files". The only way to access the content of a file from the main workspace is via the symlink hack described in the first post.

2. Some external dependencies require running a setup tool, which outputs files with non-deterministic paths. I want to inspect the log file generated by those tools to discover the paths, so they can be inserted into the generated `BUILD` file.

**Comment by jin** [comment]
+1, would be nice to have an OS-independent `cat` like feature in a repository rule.

**Comment by jin** [comment]
This came up in a discussion with @dkelmer earlier on again. Currently, to do cross-platform `cat` in a repository rule, we'll need to do this:

```
def _cat_file(repository_ctx, filepath):
    # For Windows, use cat from msys.
    cat = "C:\\msys64\\usr\\bin\\cat" if (_is_windows(repository_ctx)) else repository_ctx.which("cat")
    exec_result = repository_ctx.execute([cat, repository_ctx.path(filepath)])
    if (exec_result.return_code != 0):
        fail("Error while trying to read %s: %s" % (filepath, exec_result.stderr))
    return exec_result.stdout
```

It'll be nice to shorten this to `content = repository_ctx.read(file)`.

**Comment by cgruber** [comment]
I mean, it's that or add toolchains. :)
